---
title: "Demultiplexing - ICGC_3299"
author: "Ramon Massoni-Badosa"
date: "2/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
We had problems in 01-demultplexing.Rmd to demultiplex the cells from the patient ICGC_3299. This is because a specific blood sample (ICGC_3299_19_2768B) does not contain any counts for its corresponding hashtag oligonucleotide. This can be due to the following challenges:

- The labeling of this sample didn't work because the antibody was degraded or not function properly. In this case our dataset can have a considerable number of cells of this sample, which can challenge interpretation if we cannot assign them to their original condition properly.
- There were too few cells to tag, so our dataset has no cells which belong to this condition. In this case, we can safely proceed with our analysis.

Here, we will exclude the HTO reads of this sample and demultiplex the rest to shed light into which of these two scenarios we are in.

### Load packages

```{r}
library(Seurat)
library(Matrix)
library(readxl)
library(tidyverse)
```

```{r}
richter_list <- list()
libraries <- c("ICGC_3299_01", "ICGC_3299_02")
for (lib in libraries) {
  print(lib)
  lib_path <- str_c("data/BCLLATLAS_10/", lib, "/filtered_feature_bc_matrix/")
  expression_matrix <- readMM(str_c(lib_path, "matrix.mtx.gz"))
  barcodes <- read_csv(str_c(lib_path, "barcodes.tsv.gz"), col_names = FALSE)
  colnames(barcodes) <- "barcode"
  features <- read_tsv(str_c(lib_path, "features.tsv.gz"), col_names = FALSE)
  colnames(features) <- c("ensembl", "symbol", "feature_type")
  rownames(expression_matrix) <- features$symbol
  colnames(expression_matrix) <- barcodes$barcode
  
  # Separate HTO and RNA matrices
  hto_ind <- which(str_detect(features$feature_type, "Antibody Capture"))
  rna_ind <- which(str_detect(features$feature_type, "Gene Expression"))
  richter_hto <- expression_matrix[hto_ind, ]
  richter_rna <- expression_matrix[rna_ind, ]
  
  # Setup Seurat object
  richter <- CreateSeuratObject(counts = richter_rna)
  
  # Normalize RNA data with log normalization
  richter <- NormalizeData(richter)
  
  # Find and scale variable features
  richter <- FindVariableFeatures(richter, selection.method = "vst")
  richter <- ScaleData(richter, features = VariableFeatures(richter))
  
  # Add HTO as an independent assay
  richter_hto <- richter_hto[rownames(richter_hto) != "ICGC_3299_19_2768B", ]
  richter[["HTO"]] <- CreateAssayObject(counts = richter_hto)
  richter <- NormalizeData(richter, assay = "HTO", normalization.method = "CLR")
  
  # Demultiplex
  richter <- HTODemux(richter, assay = "HTO", positive.quantile = 0.99)
  
  # Append to list of Seurat objects
  richter_list[[lib]] <- richter
}
```

Visualization:

```{r}
ridge_l <- map(richter_list, function(richter) {
  Idents(richter) <- "HTO_maxID"
  RidgePlot(richter, assay = "HTO", features = rownames(richter[["HTO"]]))
})
ridge_l
heatmap_l <- map(richter_list, function(richter) {
  HTOHeatmap(richter, assay = "HTO", ncells = 5000)
})
heatmap_l

# Save
walk2(ridge_l, names(ridge_l), function(ridge, lib) {
  ggsave(
    filename = str_c("results/plots/", lib, "_hashtag_demux_ridge.pdf"),
    plot = ridge, height = 9, 
    width = 16
  )
})
walk2(heatmap_l, names(heatmap_l), function(heat, lib) {
  pdf(file = str_c("results/plots/", lib, "_hashtag_demux_heatmap.pdf"), height = 4, width = 7)
  print(heat)
  dev.off()
})
```

# Merge Seurat objects
To merge the 2 Seurat objects into one we use the [`merge` command](https://satijalab.org/seurat/v3.0/merge_vignette.html):

```{r}
richter_merged <- merge(
  x = richter_list$ICGC_3299_01, 
  y = richter_list$ICGC_3299_02,
  add.cell.ids = libraries, 
  project = "RICHTER"
)
barcodes_chr <- str_sub(colnames(richter_merged), start = -19, end = -1)
library <- map2_chr(colnames(richter_merged), barcodes_chr, ~ str_remove(.x, .y))
richter_merged$library <- library
```


# Assign label to each condition identifier

```{r}
condition_id <- read_excel(
  "data/BCLLATLAS_10/BCLLATLAS_10_ids-Richter Syndrom.xlsx"
)
condition_id
```


As the Seurat object has two more conditions than the dataframe (Negative and Doublet), we need to include them in the data.frame:

```{r}
condition_id$id <- str_replace_all(condition_id$id, pattern = "_", replacement = "-")
vec <- c("Doublet", "Negative")
na_vec <- c(NA, NA)
tbl <- tibble(
  id = vec, 
  description = vec, 
  Source = na_vec, 
  "ICGC Number" = na_vec, 
  "Time-post-extraction" = na_vec, 
  comments = na_vec
)
condition_id <- rbind(condition_id, tbl)
conditions <- condition_id$description
names(conditions) <- condition_id$id
conditions
```

We can change the id for the condition

```{r}
richter_merged$status <- conditions[richter_merged$hash.ID]
richter_merged$patient_id <- str_extract(colnames(richter_merged), "ICGC_3299")
batch_vec <- c("1", "2")
names(batch_vec) <- libraries
richter_merged$batch <- batch_vec[richter_merged$library]
```

Importantly, samples for every time-point of every patient where taken from a specific niche (peripheral blood or lymph node):

```{r}
source <- condition_id$Source
names(source) <- condition_id$id
richter_merged$source <- source[richter_merged$hash.ID]
```

And each blood sample is estimated to have spent some time at room temperature before cryopreservation (sampling time):

```{r}
sampling_time <- condition_id$`Time-post-extraction`
names(sampling_time) <- condition_id$id
richter_merged$sampling_time <- sampling_time[richter_merged$hash.ID]
```


# Save demultiplexed Seurat object
Finally, we can save it as .RDS for future analysis:

```{r}
selected <- c("nCount_RNA", "nFeature_RNA", "hash.ID", "library", "status", "patient_id", "batch", "source", "sampling_time")
richter_merged@meta.data <- richter_merged@meta.data[, selected]
saveRDS(richter_merged, "results/R_objects/richter_Seurat_3299_demultiplexed.rds")
```

# Session Info

```{r}
sessionInfo()
```


