---
title: "QC - normalisation - Richter - New samples"
author: "Ramon Massoni-Badosa"
date: "2/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
### KEY PARAMETERS
library(BiocStyle)

# Cell QC
min_total_counts <- 1000
max_total_counts <- 12000
min_total_genes <- 250
max_total_genes <- 3500
max_pct_mt_expression <- 25

# Gene QC
min_total_cells <- 9
```

# Introduction

The objective of this notebook is to filter-out poor-quality cells and lowly expressed and normalize UMI of the new samples we sequenced. Briefly, we profiled the single-cell transcriptomes of the clinical time-points that had fewer cells, so that our conclusions are more robust. This time we did not use cell hashing, so we will read the expression matrices directly.


# Pre-processing

## Package loading

```{r}
library(scater)
library(scran)
library(Seurat)
library(ggpubr)
library(Matrix)
library(readxl)
library(tidyverse)
```

## Source script with functions

```{r}
source("bin/utils.R")
```

## Load data

We load the demuliplexed `Seurat` object:

```{r}
# Load and process matrices
path_to_new_samples <- "data/BCLLATLAS_29"
directories <- path_to_new_samples %>%
  list.dirs() %>%
  str_subset("filtered_feature_bc_matrix")
libraries <- list.files("data/BCLLATLAS_29")
expression_matrices <- purrr::map2(directories, libraries, function(x, y) {
  features <- read_tsv(
    gzfile(str_c(x, "features.tsv.gz", sep = "/")),
    col_names = c("ensembl", "symbol", "type")
  )
  barcodes <- read_tsv(
    gzfile(str_c(x, "barcodes.tsv.gz", sep = "/")),
    col_names = "barcode"
  )
  barcodes$barcode <- str_c(y, barcodes$barcode, sep = "__")
  mat <- readMM(gzfile(str_c(x, "matrix.mtx.gz", sep = "/")))
  rownames(mat) <- features$symbol
  colnames(mat) <- barcodes$barcode
  mat
})
names(expression_matrices) <- libraries
expression_matrix <- Reduce(Matrix::cbind2, expression_matrices)


# Load and process metadata
metadata <- read_excel(
  "data/BCLLATLAS_10/BCLLATLAS_10_ids.xlsx",
  col_names = TRUE
)
```


Now we can create the Seurat object and add the metadata

```{r}
richter <- CreateSeuratObject(expression_matrix)
richter$sample <- colnames(richter) %>%
  str_split(pattern = "__") %>%
  purrr::map_chr(1)
richter$id <- case_when(
  richter$sample == "00_30" ~ "ICGC_012_00_30",
  richter$sample == "04-67" ~ "ICGC_019_04_67",
  richter$sample == "05_662" ~ "ICGC_012_05_662",
  richter$sample == "15-194" ~ "ICGC_365_15_194",
  richter$sample == "365_07" ~ "ICGC_365_365_07"
)
richter$description <- case_when(
  richter$sample == "00_30" ~ "dx",
  richter$sample == "04-67" ~ "diagnosis_sp",
  richter$sample == "05_662" ~ "progression",
  richter$sample == "15-194" ~ "richter_post-ofatu-duvelisib",
  richter$sample == "365_07" ~ "relapse1_post-R-Brenda"
)
richter$source <- case_when(
  richter$sample == "00_30" ~ "NA",
  richter$sample == "04-67" ~ "PB",
  richter$sample == "05_662" ~ "PB",
  richter$sample == "15-194" ~ "LN",
  richter$sample == "365_07" ~ "PB"
)
richter$patient_id <- case_when(
  richter$sample == "00_30" ~ "ICGC_012",
  richter$sample == "04-67" ~ "ICGC_019",
  richter$sample == "05_662" ~ "ICGC_012",
  richter$sample == "15-194" ~ "ICGC_365",
  richter$sample == "365_07" ~ "ICGC_365"
)

richter
```


# Cell QC

## Calculate QC metrics

There are 3 essential quality control metrics that will determine if we include or exclude a cell:

- Library size (total UMI): broken cells or empty droplets will have very little RNA. Likewise, doublets will have too many total UMI. We aim to discard both options.
- Number of detected genes: highly correlated with the former. Stressed cells will have an increased dropout rate, and therefore fewer detected genes.
- Percentage of mitochondrial expression: apoptotic or lysed cells will possess a high mitochondrial expression. That is, if the cell had shears in its membrane, the cytoplasmic RNA might leak outwards, but the mtRNA can remain trapped inside the cell.

The only metric missing is the mt expression:

```{r}
richter[["percent_mt"]] <- PercentageFeatureSet(richter, pattern = "^MT-")
```

## Joint distribution
An important exploratory analysis is the joint distribution between the 3 metrics. Specially, a high mitochondrial expression can also be present in metabolically active cells. Therefore, we need to assess how this covaries with total counts. If we find that most cells with large mitochondrial activity also have few genes/UMI, then we can be certain that they are of poor-quality:

```{r}
qc_titles <- c("Library Size (total UMI)", "Number of detected genes", "% mitochondrial expression")
joint_qc_gg <- richter@meta.data %>% 
  ggplot(aes(nCount_RNA, nFeature_RNA, color = percent_mt)) +
    geom_point(alpha = 0.5) +
    scale_color_viridis_c() +
    labs(x = qc_titles[1], y = qc_titles[2], color = qc_titles[3]) +
    theme_classic()
joint_qc_gg
```

Indeed, we observe how cells with few UMI/genes also harbor a high mitochondrial expression.

## Thresholds
Now that we have a better appreciation for our data, we can proceed to decide on the thresholds for each metric. We will do so by plotting a histogram for each metric:

### Library size

```{r}
qc_colnames <- c("nCount_RNA", "nFeature_RNA", "percent_mt")
hist_counts_gg1 <- plot_histogram_seurat(  
  richter, 
  qc_metric = qc_colnames[1], 
  limits = c(0, 30000),
  title = qc_titles[1],
  log_scale = FALSE
)
hist_counts_gg1
```

As there is a long tail, we need more resolution to decide on the thresholds:

```{r}
hist_counts_gg2 <- plot_histogram_seurat(  
  richter, 
  qc_metric = qc_colnames[1], 
  limits = c(0, 4000),
  title = qc_titles[1],
  log_scale = FALSE
)
hist_counts_gg2
```

We see a bimodal distribution, with the first peak representing poor-quality cells. We will retain cells that have at least `r min_total_counts` counts. Moreover, we require less than `r max_total_counts`:

```{r}
hist_counts_gg3 <- plot_histogram_seurat(  
  richter, 
  qc_metric = qc_colnames[1], 
  limits = c(0, 45000),
  title = qc_titles[1],
  log_scale = TRUE
)
hist_counts_gg3 +
  geom_vline(xintercept = min_total_counts, linetype = "dashed", color = "red") +
  geom_vline(xintercept = max_total_counts, linetype = "dashed", color = "red") 
```

### Number of detected genes

```{r}
hist_n_genes_gg1 <- plot_histogram_seurat(  
  richter, 
  qc_metric = qc_colnames[2], 
  limits = c(0, 6500),
  title = qc_titles[2],
  log_scale = FALSE
)
hist_n_genes_gg1
```

We will threshold at a minimum of `r min_total_genes` and a maximum of `r max_total_genes`:

```{r}
hist_n_genes_gg1 +
  geom_vline(xintercept = min_total_genes, linetype = "dashed", color = "red") +
  geom_vline(xintercept = max_total_genes, linetype = "dashed", color = "red") 
```

### Mitochondrial expression

```{r}
mt_hist <- plot_histogram_seurat(  
  richter, 
  qc_metric = qc_colnames[3], 
  limits = c(0, 100),
  title = qc_titles[3],
  log_scale = FALSE
)
mt_hist
```

Overall, we will consider as poor-quality any cell that satisfies any of the following conditions:

* Library size: < `r min_total_counts`, > `r max_total_counts`
* Number of detected genes: < `r min_total_genes`, > `r max_total_genes` 
* Mitochondrial expression: > `r max_pct_mt_expression`

```{r}
is_poor_quality <- 
  richter$nCount_RNA < min_total_counts |
  richter$nCount_RNA > max_total_counts |
  richter$nFeature_RNA < min_total_genes |
  richter$nFeature_RNA > max_total_genes |
  richter$percent_mt > max_pct_mt_expression
table(is_poor_quality)
```

## Cell filtering

In light of the above, we will discard the following cells:

```{r}
richter$is_low_quality <- is_poor_quality
richter <- subset(richter, subset = is_low_quality == FALSE)
richter
```

# Gene QC
Let us compute, for each gene, the number of cells in which we can detect at least 1 UMI:

```{r}
n_cells <- rowSums(as.matrix(richter[["RNA"]]@counts) > 0)
gene_qc_gg <- n_cells %>% 
  as.data.frame() %>% 
  ggplot(aes(n_cells)) + 
    geom_histogram(bins = 100, alpha = 0.75) +
    scale_x_log10("Number of cells") +
    theme_bw() 
gene_qc_gg
```


We see two peaks, the first one of which corresponds to lowly expressed genes. As explained in [Luecken MD et al.](https://www.embopress.org/doi/pdf/10.15252/msb.20188746): "a guideline to setting this threshold is to use the minimum cell cluster size that is of interest and leaving some leeway for dropout effects". As we will not rely on clusters that have fewer than `r min_total_cells` cells, we will use it as a filter:

```{r}
gene_qc_gg +
  geom_vline(xintercept = min_total_cells, color = "red", linetype = "dashed")
richter_sce <- as.SingleCellExperiment(richter)
richter_sce <- richter_sce[n_cells > min_total_cells, ]
richter_sce
```

# Normalization
To confidently compare gene expression between cells, we need to correct for two biases:

- Differences in library size: if one cell has more total counts than another due to sampling effects, genes that are equally expressed will show up as upregulated.
- Between-sample systematic biases: if two conditions/donors were sequenced at different depths or in different batches, there might be compositional biases between conditions that yield a systematic up- or down-regulation.

To correct for both, we will use the [`scran` package](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7)  which according to both a [recent](https://www.nature.com/articles/s41467-019-12266-7) and an [old](https://www.nature.com/articles/nmeth.4292) review is the most robust method for scRNA-seq data normalization:

```{r}
richter_sce <- computeSumFactors(richter_sce)
summary(sizeFactors(richter_sce))
richter_sce <- logNormCounts(richter_sce)
assays(richter_sce)
logcounts(richter_sce)[1:6, 1:6]
```

Let us assess if the size factors correlate with the library size:

```{r}
size_fact_df <- data.frame(
  lib_size = richter_sce$nCount_RNA, 
  size_fact = sizeFactors(richter_sce),
  donor = richter_sce$patient_id
)
ggplot(size_fact_df, aes(lib_size, size_fact, color = donor)) +
  geom_point() +
  labs(x = "Library Size (total UMI)", y = "Scran Size Factors") +
  theme_classic2()
lm_size_fact <- lm(size_fact ~ lib_size, data = size_fact_df)
summary(lm_size_fact)
```


# Save Seurat object

Finally, we can convert it back to seurat and save it as a compressed .RDS file for future analysis:

```{r}
richter <- as.Seurat(richter_sce)
richter
path_to_save_obj <- "/Volumes/Massoni_external/B_cell_atlas/RICHTER/current/results/R_objects/richter_new_libraries_filtered_normalized.rds"
saveRDS(richter, path_to_save_obj)
```

# Session Info

```{r}
sessionInfo()
```

